File: src/engine/pace-engine.ts
export type CheckType = 'known' | 'secret' | 'emergency' | 'gas' | 'start' | 'finish';

export interface Segment {
  distance: number;   // miles
  speed: number | null; // required avg mph, null = free/transfer
  isReset: boolean;
  isFree: boolean;
  label?: string;
  checkType?: CheckType; // type of event at the END of this segment; pace engine ignores it
  // Fixed seconds this segment adds to the key-time schedule, independent of
  // distance/speed — a scheduled pause or gas-stop wait printed on the sheet
  // (e.g. "PAUSE 21 MINS."). Distance is typically 0. Applies regardless of
  // isFree/speed: a pause has no pace to score, but the printed key times
  // downstream still assume its duration passed, so it must still count
  // against the clock (unlike a plain free/transfer segment, which
  // contributes nothing).
  holdSeconds?: number;
}

export interface RidePosition {
  segmentIndex: number;
  distanceInSegment: number; // miles into current segment
  cumulativeDistance: number; // total miles
}

export function detectSegment(segments: Segment[], cumulativeDistance: number): RidePosition {
  let remaining = cumulativeDistance;
  for (let i = 0; i < segments.length; i++) {
    if (remaining <= segments[i].distance || i === segments.length - 1) {
      return {
        segmentIndex: i,
        distanceInSegment: remaining,
        cumulativeDistance,
      };
    }
    remaining -= segments[i].distance;
  }
  return { segmentIndex: 0, distanceInSegment: 0, cumulativeDistance };
}

// Returns key time in seconds for completed scored segments before segmentIndex
export function completedKeyTime(segments: Segment[], segmentIndex: number): number {
  let keyTime = 0;
  for (let i = 0; i < segmentIndex; i++) {
    const seg = segments[i];
    if (!seg.isFree && seg.speed !== null) {
      keyTime += (seg.distance / seg.speed) * 3600;
    }
    keyTime += seg.holdSeconds ?? 0;
  }
  return keyTime;
}

// Returns key time in seconds including partial contribution of current segment
export function computeKeyTime(
  segments: Segment[],
  segmentIndex: number,
  distanceInSegment: number
): number {
  const completed = completedKeyTime(segments, segmentIndex);
  const current = segments[segmentIndex];
  const hold = current.holdSeconds ?? 0;
  if (current.isFree || current.speed === null) {
    return completed + hold;
  }
  return completed + (distanceInSegment / current.speed) * 3600 + hold;
}

// Positive = late, negative = early (seconds)
export function computeDeviation(elapsedSeconds: number, keyTimeSeconds: number): number {
  return elapsedSeconds - keyTimeSeconds;
}

export function isInFreeSegment(segments: Segment[], segmentIndex: number): boolean {
  return segments[segmentIndex]?.isFree || segments[segmentIndex]?.speed === null;
}

export function distanceToNextEvent(segments: Segment[], position: RidePosition): number | null {
  const current = segments[position.segmentIndex];
  if (!current) return null;
  return current.distance - position.distanceInSegment;
}

/**
 * Returns true if any segment boundary crossed since the last update carried
 * a reset checkpoint.
 *
 * Keying off index advance rather than proximity to the segment start means
 * a reset cannot be silently dropped because two BLE notifications bracketed
 * the boundary and the second landed more than the proximity threshold past it.
 * Also handles the case where multiple boundaries are crossed in one update —
 * the reset need not be on the final segment entered.
 */
export function crossedReset(
  segments: Segment[],
  prevSegmentIndex: number,
  currentSegmentIndex: number
): boolean {
  if (currentSegmentIndex <= prevSegmentIndex) return false;
  for (let i = prevSegmentIndex + 1; i <= currentSegmentIndex; i++) {
    if (segments[i]?.isReset) return true;
  }
  return false;
}

File: firmware/core/pace_engine.c
#include "pace_engine.h"

pe_position_t pe_detect_segment(const pe_segment_t *segments, size_t count,
                                double cumulative_distance) {
  double remaining = cumulative_distance;
  for (size_t i = 0; i < count; i++) {
    if (remaining <= segments[i].distance || i == count - 1) {
      pe_position_t pos = {
        .segment_index = (int32_t)i,
        .distance_in_segment = remaining,
        .cumulative_distance = cumulative_distance,
      };
      return pos;
    }
    remaining -= segments[i].distance;
  }
  pe_position_t pos = {0, 0.0, cumulative_distance};
  return pos;
}

double pe_completed_key_time(const pe_segment_t *segments, size_t count,
                             int32_t segment_index) {
  double key_time = 0.0;
  for (int32_t i = 0; i < segment_index && (size_t)i < count; i++) {
    const pe_segment_t *seg = &segments[i];
    if (!seg->is_free && seg->has_speed) {
      key_time += (seg->distance / seg->speed) * 3600.0;
    }
  }
  return key_time;
}

double pe_compute_key_time(const pe_segment_t *segments, size_t count,
                           int32_t segment_index, double distance_in_segment) {
  double completed = pe_completed_key_time(segments, count, segment_index);
  const pe_segment_t *current = &segments[segment_index];
  if (current->is_free || !current->has_speed) {
    return completed;
  }
  return completed + (distance_in_segment / current->speed) * 3600.0;
}

double pe_compute_deviation(double elapsed_seconds, double key_time_seconds) {
  return elapsed_seconds - key_time_seconds;
}

bool pe_is_in_free_segment(const pe_segment_t *segments, size_t count,
                           int32_t segment_index) {
  if (segment_index < 0 || (size_t)segment_index >= count) return false;
  return segments[segment_index].is_free || !segments[segment_index].has_speed;
}

bool pe_crossed_reset(const pe_segment_t *segments, size_t count,
                      int32_t prev_segment_index, int32_t current_segment_index) {
  if (current_segment_index <= prev_segment_index) return false;
  for (int32_t i = prev_segment_index + 1; i <= current_segment_index; i++) {
    if (i >= 0 && (size_t)i < count && segments[i].is_reset) return true;
  }
  return false;
}
